configs/configs.txt
#Parameters for uSim/nuSim
#Visualze the MuJoCo environment during training
visualize = False

#Print output statistics during training
verbose_training = True

###PATHS TO THE INPUT DATA/FILES###
###-------------------------------------------------------------------
## DO NOT change these paths if using the default paths for input data/files

#The path to the folder that contains the musculoskeletal model file
musculoskeletal_model_path = musculoskeletal_model/musculoskeletal_model.xml

#Path to the folder that contains intial pose (init_qpos.npy and init_qvel.npy) files
initial_pose_path = initial_pose

#Path to the folder that contains the experimental kinematics data
kinematics_path = kinematics_data

#Path to the folder that contains the experimental neural data
nusim_data_path = nusim_neural_data

#Path to the folder that contains the experimental stimulus data
stimulus_data_path = stimulus_data

### PATHS FOR SAVING THE OUTPUT / TEST DATA###
### DO NOT change these paths if using the default paths for the saved output/test data
###--------------------------------------------------
#Save the agent networks after save_iter 
save_iter = 100

#Path to the root directory
root_dir = "."

#Path to save the agent's neural networks
checkpoint_folder = "./checkpoint"

#Path for saving the statististics for training
statistics_folder = "training_statistics"

#Save name for the agent's networks
checkpoint_file = "agent_networks"

#Save name for saving the test data
test_data_filename = "test_data"

#Load the saved networks from the previous session for further training
load_saved_nets_for_training = False

### Kinematics Preprocessing Parameters
###----------------------------------------------------------------------
#Kinematics preprocessing for simulation
#Adjustment instructions:

#The timestep for the simulation: Keep 0 for default simulation timestep
sim_dt = 0   # in seconds

#The frames/timepoints for which the same action should be repeated during training of the agent
#For finer movements user smaller frame_repeat, but it will also increase the training time
frame_repeat = 5

#Number of fixedsteps in the beginning of the simulation. The target will remain at kinematic[timestep=0] for n_fixedsteps
#If a good initial position is found using CMA-ES / IK Optimization, n_fixedsteps = 25 is a good estimate. Otherwise increase
#if the starting reward does not increase with the training iterations.
n_fixedsteps = 25

#Timestep limit is max number of timesteps after which the episode will terminate.
#Multiple cycles of the same condition will be simulated if the timestep_limit > number of timsteps for that condition.
timestep_limit = 8000

#Adjusts/scales the length of the trajectory
#Should be the same as num_markers/targets
trajectory_scaling = [26.3157894737]

#Adjusts the starting point of the kinematics trajectory
#Should be the same as num_markers/targets, num_coords=3
center = [[0.06, 0.083, 0]]

###-----------------------------------------------------------------------------------

###Sensory Feedback Processing Parameters --------------------------------------------
#Specifies the sensory feedback to the agent/network
#True, if this feedback should be included in state feedback to the agent's network/controller
#False, if this feedback should not be included in the state feedback to the agent's network/controller

#Stimulus feedback consists of provided experimental stimulus data
stimulus_feedback = False

#Proprioceptive feedback consists of muscle lengths and velocities
proprioceptive_feedback = True

#Muscle forces consist of appled muscle forces 
muscle_forces = False

#Joint feedback consists of joint positions and velocities
joint_feedback = False

#Visual feedback consists of x/y/z coordinates of the specified bodies in the model
#If visual_feedback is True, specify the names of the bodies from musculoskeletal_model.xml for which the feedback should be included
visual_feedback = False

#Append the musculo bodies from which visual feedback should be included
#This list can also consist of targets/markers
#Append targetn-1 for visual feedback from targets/markers in the kinematics.pkl file
#'target0' corresponds to the visual feedback from the first target/marker, target1 to the second target/marker and so on
visual_feedback_bodies = [hand, target0]

#Specify the names of the bodies as tuples(separated by ; with no spaces) for which the visual distance should be included in the feedback
#Leave blank if the visual distance is not to be included in the feedback
#Visual distance between the bodies will be included
#e.g visual_distance_bodies = [[hand;target0], [elbow;target0]] will include the distance between the hand/elbow and first marker in sensory feedback
visual_distance_bodies = [[hand;target0]]

#Specify the names of the bodies for which the visual velocity should be included in the feedback
#Leave blank if the visual velocity is not to be included in the feedback
#Appends the absolute musculo body velocity, e.g. visual_velocity = [hand, target0]
#will include the xyz velocities of hand and target0
visual_velocity = []

#Specify the delay in the sensory feedback in terms of the timepoints
sensory_delay_timepoints = 0

### -----------------------------------------------------------------------------------
###Specifications for Regularizations with the policy network
#Specify the weighting with various neural regularizations used in uSim/nuSim

#weighting with loss for enforcing simple neural dynamics for uSim/nuSim
alpha_usim = 0.1

#weighting with loss for minimizing the neural activations for uSim/nuSim
beta_usim = 0.01

#weighting with loss for minimizing the synaptic weights for uSim/nuSim
gamma_usim = 0.001

#weighting with loss for nuSim constraining a sub-population of RNN units to experimentally recorded neurons for nuSim
zeta_nusim = 0

### --------------------------------------------------------------------------------------

### SAC TRAINING ###

#The neural network model to use in the agent, can be ['rnn', 'gru']
model = rnn

#The number of hidden units in the layers of the agent's neural network
hidden_size = 256

#The mode of simulation can be [train, test, SFE, sensory_pert, neural_pert, musculo_properties]
mode = "train"

#DRL specific parameters.
gamma = 0.99
tau = 0.005
lr = 0.0003
alpha = 0.20
automatic_entropy_tuning = True
seed = 123456
policy_batch_size = 8
policy_replay_size = 4000
multi_policy_loss = True
batch_iters = 1
total_episodes = 1000000
condition_selection_strategy = "reward"

# The RL algorithm be [SAC, DDPG, TD3]
RL_algorithm = 'SAC'

# For the TD3 algorithm, the following parameters must be specified
target_noise = .2
target_noise_clip = .5
policy_delay = 2

cuda = True

configs/monkey_configs.txt
### MUSCLE MODEL ###
musculoskeletal_model_path = musculoskeletal_model/musculoskeletal_model.xml
initial_pose_path = initial_pose
kinematics_path = kinematics_data
nusim_data_path = nusim_neural_data

### SAC TRAINING ###
model = rnn
gamma = 0.99
tau = 0.005
lr = 0.0003
alpha = 0.20
automatic_entropy_tuning = True
seed = 123456
policy_batch_size = 8
hidden_size = 256
policy_replay_size = 4000
multi_policy_loss = True
batch_iters = 1
total_episodes = 1000000
mode = "train"
condition_selection_strategy = "reward"
cuda = True

### SAVING ###
save_iter = 100
root_dir = "."
checkpoint_folder = "checkpoint"
statistics_folder = "training_statistics"
checkpoint_file = "agent_networks"
test_data_filename = "test_data"
load_saved_nets_for_training = "False"

config.py
import configargparse

def list_of_list_of_floats(x):
    temp = x.replace('[', '').replace(']', '').split(',')
    temp2 = [float(temp_e) for temp_e in temp]

    return temp2

def list_of_string_names(x):

    return x

def list_of_tuples_of_strings(x):

    temp = x.replace('[', '').replace(']', '').split(';')

    return temp

def boolean_string(s):
    if s not in {'False', 'True'}:
        raise ValueError('Not a valid boolean string')
    return s == 'True'

def config_parser():
    parser = configargparse.ArgumentParser()
    
    parser.add_argument("--config", is_config_file=True, help="config file path")
    
    parser.add_argument('--model', 
                        type=str, 
                        default="rnn",
                        help='rnn, gru')

    parser.add_argument('--gamma', 
                        type=float, 
                        default=0.99, 
                        help='discount factor for reward (default: 0.99)')

    parser.add_argument('--tau', 
                        type=float, 
                        default=0.005, 
                        help='target smoothing coefficient(τ) (default: 0.005)')

    parser.add_argument('--lr', 
                        type=float, 
                        default=0.0003, 
                        help='learning rate (default: 0.001)')

    parser.add_argument('--alpha', 
                        type=float, 
                        default=0.2, 
                        help='Temperature parameter α determines the relative importance of the entropy\
                                term against the reward (default: 0.2)')

    parser.add_argument('--automatic_entropy_tuning', 
                        type=boolean_string, 
                        default=True, 
                        help='Automaically adjust α (default: False)')

    parser.add_argument('--seed', 
                        type=int, 
                        default=123456, 
                        help='random seed (default: 123456)')

    parser.add_argument('--policy_batch_size', 
                        type=int, 
                        default=8, 
                        help='batch size (default: 8)')

    parser.add_argument('--hidden_size', 
                        type=int, 
                        default=256, 
                        help='hidden size (default: 1000)')

    parser.add_argument('--policy_replay_size', 
                        type=int, 
                        default=50000, 
                        help='size of replay buffer (default: 2800)')

    parser.add_argument('--multi_policy_loss', 
                        type=boolean_string, 
                        default=False, 
                        help='use additional policy losses')

    parser.add_argument('--batch_iters', 
                        type=int, 
                        default=1, 
                        help='iterations to apply update')

    parser.add_argument('--cuda', 
                        action="store_true",
                        help='run on CUDA (default: False)')

    parser.add_argument('--visualize', 
                        type=boolean_string, 
                        default=False,
                        help='visualize monkey/mouse')

    parser.add_argument('--root_dir', 
                        type=str, 
                        default='',
                        help='specify you root directory')

    parser.add_argument('--checkpoint_file', 
                        type=str, 
                        default='agent_networks',
                        help='specify the name of the file in which you would like to save model weights/training params (do not add extension). Also saves statistics file in root of project folder with same name')

    parser.add_argument('--checkpoint_folder', 
                        type=str, 
                        default= 'checkpoint',
                        help='specify the name of the folder in which you would like to save the checkpoint file')

    parser.add_argument('--statistics_folder', 
                        type=str, 
                        default= 'training_statistics',
                        help='specify the name of the folder in which you would like to save the training statistics')

    parser.add_argument('--total_episodes', 
                        type=int, 
                        default=5000000, 
                        help='total number of episodes')

    parser.add_argument('--save_iter', 
                        type=int, 
                        default=100, 
                        help='number of episodes until checkpoint is saved')

    parser.add_argument('--mode', 
                        type=str, 
                        default="train", 
                        help='select whether to train or test a model (train, test, SFE, sensory_pert, neural_pert, musculo_properties)')


    parser.add_argument('--verbose_training', 
                        type=boolean_string, 
                        default=False, 
                        help='Print statistics during training')

    parser.add_argument('--load_saved_nets_for_training', 
                        type=boolean_string, 
                        default=False,
                        help='select whether to train or test a model (train, test)')

    parser.add_argument('--musculoskeletal_model_path', 
                        type=str, 
                        default='musculoskeletal_model/musculoskeletal_model.xml',
                        help='path of musculoskeletal model')

    parser.add_argument('--initial_pose_path', 
                        type=str, 
                        default='initial_pose',
                        help='path of musculoskeletal model')

    parser.add_argument('--kinematics_path', 
                        type=str, 
                        default='kinematics_data',
                        help='path to kinematics data')

    parser.add_argument('--nusim_data_path', 
                        type=str, 
                        default='nusim_neural_data',
                        help='path to nusim neural data for training and testing')

    parser.add_argument('--stimulus_data_path', 
                        type=str, 
                        default='stimulus_data',
                        help='path to experimental stimulus data for training and testing')

    parser.add_argument('--test_data_filename', 
                        type=str, 
                        default='test_data',
                        help='filename for saving the testing data')
                        
    parser.add_argument('--condition_selection_strategy', 
                        type=str, 
                        default='reward',
                        help='whether to select the next condition based on the corresponding average reward accumulated so far')

    parser.add_argument('--sim_dt', 
                        type=int, 
                        default=0,
                        help='The timestep for the simulation: Keep 0 for default simulation timestep')

    parser.add_argument('--frame_repeat', 
                        type=int, 
                        default=5,
                        help='The frames/timepoints for which the same action should be repeated during training of the agent')

    parser.add_argument('--n_fixedsteps', 
                        type=int, 
                        default=25,
                        help='The target will remain at kinematic[timestep=0] for n_fixedsteps')

    parser.add_argument('--timestep_limit', 
                        type=int, 
                        default=1000,
                        help='Timestep limit is max number of timesteps after which the episode will terminate.')

    parser.add_argument('--trajectory_scaling', 
                        type= float, 
                        nargs= '+',
                        default= None,
                        help='Adjusts/scales the length of the trajectory')

    parser.add_argument('--center', 
                        type= list_of_list_of_floats,
                        nargs= '+',
                        default= None,
                        help='Adjusts the starting point of the kinematics trajectory')

    parser.add_argument('--stimulus_feedback', 
                        type= boolean_string,
                        default= False,
                        help='Experimental stimulus feedback to be included in the sensory feedback')

    parser.add_argument('--proprioceptive_feedback', 
                        type= boolean_string,
                        default= True,
                        help='Proprioceptive feedback consists of muscle lengths and velocities')

    parser.add_argument('--muscle_forces', 
                        type= boolean_string,
                        default= False,
                        help='Muscle forces consist of appled muscle forces')

    parser.add_argument('--joint_feedback', 
                        type= boolean_string,
                        default= False,
                        help='Joint feedback consists of joint positions and velocities')

    parser.add_argument('--visual_feedback', 
                        type= boolean_string,
                        default= False,
                        help='Visual feedback consists of x/y/z coordinates of the specified bodies in the model')

    parser.add_argument('--visual_feedback_bodies', 
                        type= list_of_string_names,
                        nargs= '*',
                        default= None,
                        help='Append the names musculo bodies from which visual feedback should be included')

    parser.add_argument('--visual_distance_bodies', 
                        type= list_of_tuples_of_strings,
                        nargs= '*',
                        default= None,
                        help='Specify the names of the bodies as tuples for which the visual distance should be included in the feedback')

    parser.add_argument('--visual_velocity', 
                        type= list_of_string_names,
                        nargs= '*',
                        default= None,
                        help='Specify the names of the bodies for which the visual velocity should be included in the feedback')

    parser.add_argument('--sensory_delay_timepoints', 
                        type= int,
                        default= 0,
                        help='Specify the delay in the sensory feedback in terms of the timepoints')

    parser.add_argument('--alpha_usim', 
                        type= float,
                        default= 0.1,
                        help='weighting with loss for enforcing simple neural dynamics for uSim/nuSim')

    parser.add_argument('--beta_usim', 
                        type= float,
                        default= 0.01,
                        help='weighting with loss for minimizing the neural activations for uSim/nuSim')

    parser.add_argument('--gamma_usim', 
                        type= float,
                        default= 0.001,
                        help='weighting with loss for minimizing the synaptic weights for uSim/nuSim')

    parser.add_argument('--zeta_nusim', 
                        type= float,
                        default= 0,
                        help='weighting with loss for nuSim constraining a sub-population of RNN units to experimentally recorded neurons for nuSim')

    parser.add_argument('--RL_algorithm',
                        type=str,
                        default="SAC",
                        help='SAC, DDPG, TD3')

    parser.add_argument('--target_noise',
                        type=float,
                        default=.2,
                        help='Standard deviation of noise added to target policy (in TD3)')

    parser.add_argument('--target_noise_clip',
                        type=float,
                        default=.5,
                        help='Maximum absolute value of noise added to target policy in (TD3)')

    parser.add_argument('--policy_delay',
                        type=int,
                        default=2,
                        help='Frequency with which the policy network is trained (in TD3)')

    return parser
